### PR TITLE
Fix Fatal error received while updating WP 5.8.2 to WP 5.9.

### PR DESCRIPTION
**Pull request by costdev**
Trac ticket: https://core.trac.wordpress.org/ticket/54546

### CODE REVIEW

**Comment by SergeyBiryukov** [comment]
Thanks for the PR! Just to clarify, this would only be needed for 5.8 and older branches, but not for trunk, right?

Should we also add a comment to explain this line?

**Comment by costdev** [comment]
> Thanks for the PR! Just to clarify, this would only be needed for 5.8 and older branches, but not for trunk, right?

This line will be needed for every branch of WordPress that uses the Requests library (both past and future branches).

Why?
- Every `< 5.9.0` branch can upgrade to `> 5.9.0` and will fatal on cURL timeout due to a missing `Requests_Exception` class.
- Every `> 5.9.0` branch will be able to rollback to `< 5.9.0` and will fatal on cURL timeout due to a missing `WpOrg\Requests\Exception` class.

This line will only need to exist in the upgrade packages.

Why?
- The upgrade package's `wp-admin/includes/update-core.php` file is the one that runs during the upgrade process.
- This will provide the fix to all existing installs upon upgrade, negating the need to force a maintenance update on existing installs.

Upgrading to `5.9.0 Beta 1` and rolling back to `5.8.2` has been tested by myself and [@mkaz](https://github.com/mkaz).
This test used a custom upgrade package with the line added to its `wp-admin/includes/update-core.php` file.

> Should we also add a comment to explain this line?

Absolutely. How does this sound?

```php
/**
 * Namespacing was introduced in Requests 2.0.
 * 
 * This file MUST be loaded into memory before the upgrade process
 * to prevent a fatal error on cURL timeout due to class names that
 * no longer exist after the upgrade.
 * 
 * This applies to all WordPress versions that use the Requests library.
 */
require_once ABSPATH . 'wp-includes/Requests/Exception.php';
```

**Comment by hellofromtonya** [comment]
>This line will be needed for every branch of WordPress that uses the Requests library (both past and future branches).

@costdev The upgrader uses the "to" version `wp-admin/includes/update-core.php` file, meaning only 5.9 needs this code. For older WordPress versions upgrading to 5.9, this PR will ensure that missing "from" version exception file is loaded into memory from the older WordPress version.

Why then does this need to be backported to all the older WordPress branches?

File: src/wp-admin/includes/update-core.php [hunk 1]
```diff
@@ -18,6 +18,8 @@
  */
 global $_old_files;
 
+require_once ABSPATH . 'wp-includes/Requests/Exception.php';
```
Comment (hellofromtonya) [inline src/wp-admin/includes/update-core.php]: Is this the _only_ missing "from" version file in memory that might be needed during the upgrade process? It's the one directly involved in the ticket. But have we confirmed this is the only one that is needed? 

For example, there are other types of exceptions in the Requests library. See the `wp-includes/Requests/Exception/` folder on [5.5](https://github.com/WordPress/wordpress-develop/tree/5.5/src/wp-includes/Requests/Exception), [5.6](https://github.com/WordPress/wordpress-develop/tree/5.6/src/wp-includes/Requests/Exception), [5.7](https://github.com/WordPress/wordpress-develop/tree/5.7/src/wp-includes/Requests/Exception) and [5.8](https://github.com/WordPress/wordpress-develop/tree/5.8/src/wp-includes/Requests/Exception).

May want to explore what files in the HTTP and Requests library are used and then load them here.

Comment (costdev) [reply src/wp-admin/includes/update-core.php]: > > This line will be needed for every branch of WordPress that uses the Requests library (both past and future branches).
> 
> @costdev The upgrader uses the "to" version `wp-admin/includes/update-core.php` file, meaning only 5.9 needs this code. For older WordPress versions upgrading to 5.9, this PR will ensure that missing "from" version exception file is loaded into memory from the older WordPress version.
> 
> Why then does this need to be backported to all the older branches?

"to" is only 5.9 when upgrading. When rolling back, "to" becomes the earlier version, and the fatal is thrown on missing `WpOrg\Requests\Exception`.

**Comment by hellofromtonya** [comment]
>The upgrader uses the "to" version `wp-admin/includes/update-core.php` file, meaning only 5.9 needs this code. For older WordPress versions upgrading to 5.9, this PR will ensure that missing "from" version exception file is loaded into memory from the older WordPress version.
>
>Why then does this need to be backported to all the older WordPress branches?

Sharing a conversation with @costdev.

The Upgrader does load and use "to" version's `wp-admin/includes/update-core.php` file. This means upgrading from an older version of WordPress ("from" version) to 5.9 ("to" version) is handled by 5.9.

Back downgrading (rolling back) to an older version from 5.9 will cause a fatal error for the same reason as upgrading to 5.9. So the backports are required to ensure the older ("to") version's `wp-admin/includes/update-core.php` file also loads the needed file(s) into memory to avoid the fatal error.

Backporting does not require a point release. Why? The Upgrader fetches the "to" version from wp.org. Backporting then upgrades the upgrader package (wow that was a clumpsy way to phrase it) for use when rolling back from WordPress 5.9.

File: src/wp-admin/includes/update-core.php (hunk 1 above)
Comment (hellofromtonya) [reply src/wp-admin/includes/update-core.php]: Another idea is to load all of the Requests files into memory. That way all are available for any edge case or exception that might happen.

**Comment by hellofromtonya** [comment]
Update:
@costdev and I experimented with creating a "preloader" to load all "from" version files into memory (including the entire Requests library). We have a working model. There are pain points to resolve in downgrading from 5.9.  Plus a preloader introduces complexity that needs extensive testing to avoid introducing unexpected upgrade issues. Given both of these points, it's too late in this beta period to introduce a "preloader".

Instead, let's first preload the file that directly caused the reported fatal error. Then continue monitoring feedback for other files that _may_ also need preloading.

A "preloader" can be explored during 5.9 if the entire Requests library does need preloading. Else, it can be explored in a future release.

**Review by hellofromtonya — APPROVED** [review]
Let's commit this patch as it preloads the file that is directly causing the reported fatal error.

**Comment by pbiron** [comment]
@hellofromtonya see https://github.com/WordPress/wordpress-develop/pull/1999#issuecomment-985265243.  I think it would be good to add that inline comment (or something like it) before committing.

File: src/wp-admin/includes/update-core.php (hunk 1 above)
Comment (hellofromtonya) [inline src/wp-admin/includes/update-core.php]:
```suggestion
/**
 * Preload the "from" version file(s) into memory before the file(s)
 * is(are) upgraded.
 * 
 * Why? Requests 2.0.0 (introduced in WordPress 5.9.0) introduced
 * namespaces. If a class is needed after its file is upgraded but
 * it is not in memory, a fatal error is thrown.
 * Why? As its file was upgraded, the fully-qualified name of the
 * needed class is no longer available in the filesystem.
 * 
 * @since 5.9.0
 */
require_once ABSPATH . 'wp-includes/Requests/Exception.php';
```

This explanation allows for adding more files to preload.

Comment (hellofromtonya) [reply src/wp-admin/includes/update-core.php]: @costdev @pbiron what do you think of this explanation 👆? Attempting to make it clear about why files need to be preloaded vs. making it singularly specific to the one file for this ticket. Why? To give flexibility should other files also need to preloaded.

Comment (pbiron) [reply src/wp-admin/includes/update-core.php]: how about this slightly different wording:

```
/**
 * Preload files from the current version into memory that may be needed
 * later during the update process, and may not be available after files
 * from the updated version have been installed.
 *
 * Why? Requests 2.0.0 (introduced in WordPress 5.9.0) introduced
 * namespaces. If a class is needed after its file is upgraded but
 * the class is not already available in memory, a fatal error is thrown
 * because the fully-qualified name of the needed class is no longer
 * available in the filesystem.
 *
 * @since 5.9.0
 */
```

I think it's a little easier to read.

The other suggestion I'd have is to put this comment and the `require_once` above the comment that comes before `global $_old_files;`

Comment (hellofromtonya) [inline src/wp-admin/includes/update-core.php]:
```suggestion
/**
 * Preload files from the current version into memory that may be
 * needed during the update process, and may not be available
 * after files from the updated version have been installed.
 *
 * Why? Requests 2.0.0 (introduced in WordPress 5.9.0) introduced
 * namespaces. If a class is needed after its file is upgraded but
 * the class is not already available in memory, a fatal error is
 * thrown because the fully-qualified name of the needed class
 * is no longer available in the filesystem.
 *
 * @since 5.9.0
 */
require_once ABSPATH . 'wp-includes/Requests/Exception.php';

global $_old_files;
```

**Review by pbiron — APPROVED** [review]

File: src/wp-admin/includes/update-core.php (hunk 1 above)
Comment (hellofromtonya) [reply src/wp-admin/includes/update-core.php]: Even better. Thanks @pbiron 

Done in https://github.com/WordPress/wordpress-develop/pull/1999/commits/a6279e6cfaeba04e6143747cb5100078e120a6f8 and https://github.com/WordPress/wordpress-develop/pull/1999/commits/857491c729d16f1c1cd1609859f799985c337561

File: src/wp-admin/includes/update-core.php [hunk 2]
```diff
@@ -7,6 +7,21 @@
...
+/**
+ * Preload files from the current version into memory that may be
+ * needed during the update process, and may not be available
+ * after files from the updated version have been installed.
+ *
+ * Why? Requests 2.0.0 (introduced in WordPress 5.9.0) introduced
+ * namespaces. If a class is needed after its file is upgraded but
+ * the class is not already available in memory, a fatal error is
+ * thrown because the fully-qualified name of the needed class
+ * is no longer available in the filesystem.
```
Comment (hellofromtonya) [inline src/wp-admin/includes/update-core.php]:
```suggestion
```

Removing the Requests explanation as preloading files into memory may not be limited to Requests (now and/or in the future). Having this specific comment _may_ limit focus on the intent of identifying and preloading files that may be needed before those files are upgraded.

File: src/wp-admin/includes/update-core.php [hunk 3]
```diff
@@ -7,6 +7,15 @@
...
  */
 
+/**
+ * Preload files from the current version into memory that may be
+ * needed during the update process, and may not be available
+ * after files from the updated version have been installed.
+ *
+ * @since 5.9.0
+ */
+require_once ABSPATH . 'wp-includes/Requests/Exception.php';
```
Comment (hellofromtonya) [inline src/wp-admin/includes/update-core.php]: Requests library was introduced in WordPress 4.6. Upgrading from < 4.6 would result in a fatal error.

```suggestion
global $wp_version;
if ( version_compare( $wp_version, '4.6', '>=' ) ) {
	require_once ABSPATH . WPINC . '/Requests/Exception.php';
}
```
Adding a version check here provides a guard and grouping for any other files.

Comment (hellofromtonya) [reply src/wp-admin/includes/update-core.php]: Q: Is `$wp_version` still set to the current ("from") version at this point?

A: Loads the current ("from") version `wp-includes/version.php` file at the start of `Core_Upgrader::upgrade()`.

Comment (pbiron) [reply src/wp-admin/includes/update-core.php]: good catch on the earliest version Requests lib was introduced in

**Review by SergeyBiryukov — APPROVED** [review]

**Comment by SergeyBiryukov** [comment]
> The Upgrader does load and use "to" version's `wp-admin/includes/update-core.php` file. This means upgrading from an older version of WordPress ("from" version) to 5.9 ("to" version) is handled by 5.9.
> 
> Back downgrading (rolling back) to an older version from 5.9 will cause a fatal error for the same reason as upgrading to 5.9. So the backports are required to ensure the older ("to") version's `wp-admin/includes/update-core.php` file also loads the needed file(s) into memory to avoid the fatal error.
> 
> Backporting does not require a point release. Why? The Upgrader fetches the "to" version from wp.org. Backporting then upgrades the upgrader package (wow that was a clumpsy way to phrase it) for use when rolling back from WordPress 5.9.

I'm not quite sure about the latest part. Is it about a manual downgrade or an automated rollback if the background update didn't work? Isn't the rollback package only updated once a new minor version is released?

**Comment by costdev** [comment]
I've just pushed an update that should cover downgrades without needing to backport the `require_once` line.

TL;DR - The newer WordPress version applies the fix in both of the necessary locations and the older WordPress version doesn't need to do anything.

- When upgrading: `wp-admin/includes/update-core.php` is owned by the "to" version.
- When downgrading: `wp-admin/includes/class-wp-upgrader.php` is owned by the "from" version.

Therefore:
- Upgrading from 5.8.4 to 6.1.0 will use 6.1.0's `wp-admin/includes/update-core.php` file, which will load 5.8.4's `Requests_Exception` class into memory before overwriting its contents with 6.1.0's `WpOrg\Requests\Exception` class.
- Downgrading from 6.1.0 to 5.8.4 will use 6.1.0's `wp-admin/includes/class-wp-upgrader.php` file, which will load 6.1.0's `WpOrg\Requests\Exception` class into memory before overwriting its contents with 5.8.4's `Requests_Exception` class.
- The "downgrading" portion is only ever going to be done from a version of WordPress that includes Requests, so it doesn't need a version check.
- The "upgrading" portion could be upgrading from any earlier version, so it needs to keep the version check already included in `wp-admin/includes/update-core.php`.

Happy to answer any questions to clarify anything about this update.
